src/main/scala/firrtl/FirrtlExecutionOptions.scala

```diff
@@ -0,0 +1,752 @@
+
+/** Additional arguments
+  *  - maps to [[FirrtlExecutionOptions.programArgs]]
+  *  - set with any trailing option on the command line
+  * @param value one [[scala.String]] argument
+  */
+case class ProgramArgsAnnotation(value: String) extends SingleStringAnnotation with FirrtlOption
+
+/** An explicit input FIRRTL file to read
+  *  - maps to [[FirrtlExecutionOptions.inputFileNameOverride]]
+  *  - set with `-i/--input-file`
+  *  - If unset, an [[InputFileAnnotation]] with the default input file __will not be generated__
+  * @param value input filename
+  */
+case class InputFileAnnotation(value: String) extends SingleStringAnnotation with FirrtlOption
+
+/** An explicit output file the emitter will write to
+  *   - maps to [[FirrtlExecutionOptions.outputFileNameOverride]]
+  *   - set with `-o/--output-file`
+  *  @param value output filename
+  */
+case class OutputFileAnnotation(value: String) extends SingleStringAnnotation with FirrtlOption
+
+/** An explicit output _annotation_ file to write to
+  *  - maps to [[FirrtlExecutionOptions.outputAnnotationFileName]]
+  *  - set with `-foaf/--output-annotation-file`
+  * @param value output annotation filename
+  */
+case class OutputAnnotationFileAnnotation(value: String) extends SingleStringAnnotation with FirrtlOption
+
+/** Sets the info mode style
+  *  - maps to [[FirrtlExecutionOptions.infoModeName]]
+  *  - set with `--info-mode`
+  * @param value info mode name
+  */
+case class InfoModeAnnotation(value: String = "append") extends SingleStringAnnotation with FirrtlOption
+
+/** Holds a [[scala.String]] containing FIRRTL source to read as input
+  *  - maps to [[FirrtlExecutionOptions.firrtlSource]]
+  *  - set with `--firrtl-source`
+  * @param value FIRRTL source as a [[scala.String]]
+  */
+case class FirrtlSourceAnnotation(value: String) extends SingleStringAnnotation with FirrtlOption
+
+/** Holds a filename containing one or more [[annotations.Annotation]] to be read
+  *  - this is not stored in [[FirrtlExecutionOptions]]
+  *  - set with `-faf/--annotation-file`
+  * @param value input annotation filename
+  */
+case class InputAnnotationFileAnnotation(value: String) extends SingleStringAnnotation with FirrtlOption
+
+/** Holds the name of the compiler to run
+  *  - maps to [[FirrtlExecutionOptions.compilerName]]
+  *  - set with `-X/--compiler`
+  *  - If unset, a [[CompilerNameAnnotation]] with the default compiler ("verilog") __will be generated__
+  * @param value compiler name
+  */
+case class CompilerNameAnnotation(value: String = "verilog") extends SingleStringAnnotation with FirrtlOption
+
+/** Holds the unambiguous class name of a [[Transform]] to run
+  *  - will be append to [[FirrtlExecutionOptions.customTransforms]]
+  *  - set with `-fct/--custom-transforms`
+  * @param value the full class name of the transform
+  */
+case class RunFirrtlTransformAnnotation(value: String) extends SingleStringAnnotation with FirrtlOption
```

Comment:
@azidar and I previously talked about adding this (I think I even PRed it at one point). One issue we were concerned about was the misconception that a transform could add this annotation to the AnnotationSeq during normal execution. The writer of the transform is obviously expected the transform the specified in the RunFirrtlTransformAnnotation would be run but it wouldn't be. 

That being said, maybe we can solve that issue with a `VerifyAnnotations` check or something that would check all kinds of things including that there are no annotations like this, throwing a warning or error if they're found. We could also just document that and assume that when people see their transform not being run they will check the ScalaDoc or whatever and realize that's not how it works.

Reply:
I see what you're saying here. Ugh... this gets very meta very quickly...

Using this in a transform would be an anti-pattern. If a transform needs to be composed of a sequence of transforms, then it should be calling that transform (or it's assumedly associated pass) directly as opposed to using this to run a transform.

`VerifyAnnotations` makes sense. We can talk about this... I've been thinking that we may want to divide the annotations into (at least) two groups. Certain annotations are allowed to be added at specific times. Specifically, I was toying with a division of "annotations that can be added by command lines" and "annotations which can show up in a file". I was thinking about this with the  `InputAnnotationFileAnnotation` and considering disallowing that from adding command line options annotations. That seems like a recipe for disaster as the interactions there are not clear, things can get overwritten, etc. (and it avoids having to deal with input annotation files including other input annotation files).

An alternative, which I'm not in favor of, is having the compiler try to clean up any new `RunFirrtlTransformAnnotation` by scheduling new transforms during lowering. I'm against this as it makes it very non-deterministic how the compiler is going to be run and there's a clear good pattern of having a transform schedule other transforms internally.

Reply:
One addendum, while not nearly as critical, `TargetDirAnnotation` has the same oddity, though I doubt a user would be doing this with a Transform. If a transform adds an annotation that is clearly some top-level construct like `TargetDirAnnotation`, how should we handle it? I'm very much in favor of doing nothing and possibly barfing. The command line options are, through this transform, intended to be immutable. Any options-derived annotations similarly are disallowed from being emitted by Transforms. Summarily, `FirrtlOption` really, really needs to be `sealed`.